rfcs/0089-collect-non-source-package-meta.md

```diff
@@ -0,0 +1,103 @@
+---
+feature: collect-non-source-package-meta
+start-date: 2021-03-14
+author: Robert Scott
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Collect and maintain a new `meta` attribute in packages allowing users to easily
+identify and manage their preference for binary (more broadly "non-source")
+packages.
+
+# Motivation
+[motivation]: #motivation
+
+Different users have different expectations from a software distribution. We
+acknowledge that much with the collection of license information and the
+existence of the `allowUnfree` nixpkgs option, much as Debian maintains a
+separate `-nonfree` repository.
+
+Similarly, there are a number of different reasons users may have to disfavour
+those packages not built-from-source:
+
+- Transparency: an ever-growing concern with more focus than ever on
+  supply-chain attacks.
+- Malleability: being able to conveniently override packages with patches or an
+  altered build process is a key advantage of Nix, and for nixpkgs maintainers
+  it's not generally possible to backport security fixes to binary packages.
+
+For some users, these concerns are enough to deter them from using Nix entirely.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new `meta` attribute to non-source-built packages, `fromSource = false`.
```

Comment:
This doesn't seem flexible enough:
* Along the lines of the `unfreeRedistributableFirmware` license, there should be a way to label binary packages used in bootstrapping as such. Most users who don't want binary packages will be okay with their use in bootstrapping.
* Binary fonts (mentioned below) or graphics might also be acceptable.
Do you think `allowNonSourcePredicate` is enough for the latter?

Reply:
Well, this hints slightly towards making it non-boolean, but if so, what should the options be? I think I considered finer classifications but saw it as something this could evolve into if/once we come to understand the problem better.

Reply:
A list of "types"? Because a package might contain different types!

```
{
  meta.sourceProvenance = with lib.sourceTypes; [
    sourceCode
    sourceFirmware
  ];
}
```

For an hypothetical driver with a code part, and a binary firmware part.


```
{
  meta.sourceProvenance = with lib.sourceTypes; [
    sourcePrebuiltFont
  ];
}
```

For a pre-built font.


(The naming here is clunky, I hope if gets the point across.)

Reply:
I think a list of types is usually too complex. Hence I suggest usually not using them, even though some packages might require them.
We sometimes have lists of licenses, though without clear semantics. For this attribute, a list should be interpreted as being built from *all* of those types combined.

I would call the attribute `meta.builtFrom`.
A normal `xyz-bin` package would, even if parts of it are built from source, have
```
meta.builtFrom = lib.builtFrom.binary
```
A binary font could have
```
meta.builtFrom = lib.builtFrom.binaryAsset
```
And a bootstrapping package would have
```
meta.builtFrom = lib.builtFrom.binaryBootstrap
```
The default would be
```
meta.builtFrom = lib.builtFrom.source
```

Reply:
While I agree that they are more complex, I don't think it's an issue.

Otherwise we're pushing the complexity into pre-declaring all possible combinations as part of the library representing the different types.

We need granularity here. Otherwise we'll have issues deciding which descriptor to use, and end up using the wrong generic "it's a binary", which will make filtering against or for a specific descriptor needlessly harder.

Reply:
We can do something "weird" and have `lib.builtFrom.source = ["source"]` so that you can easily do one or `++` two together. Or maybe better is `lib.buildFrom.source = {source = true;}` and then you can `//` them together but it is a bit easier for a predicate to filter the ones that you are okay with (and removes the irrelevant ordering).

Also I think instead of `builtFrom` it probably makes sense to call it `nonSourceComponents` and list the types of things that were not built from source where "types" could be things like `code`, `assets`, `docs` and similar.

I see this is somewhat against the "Why not isBinary?" below but I think it kinda agrees. `nonSourceComponents = {}` is the "purest" form and well understood. Then you just have to document the deviations.

Reply:
I really like samueldr's idea here.  Another benefit of it would be that I'd be able to say "I don't want proprietary software on my computer, but it's okay if data files are CC-BY-ND" or something.

Reply:
It's nice, my worry is just what we'd _actually_ end up with given humans being humans, most people probably not _that_ invested in understanding a complex tagging scheme well enough to perfectly represent the situation for a particular package.

This concern probably comes from my background in openstreetmap and the long, heated discussions that take place on the "tagging" mailing list debating a scheme that can perfectly represent most every situation, yet which ends up bearing little relation to what people _actually_ map with in reality, because it's too complex and verbose.

I certainly think it's important to make the common cases have very concise representations, and also allow both coarse and fine granularity. If only fine granularities are allowed, it will deter most people from bothering to add an annotation at all. Coarse data is better than nothing and can always be refined.

Reply:
I tend to agree with @risicle and @dotlambda here, in that I think we should at least start with a rougher schema, and wait until we've seen if we actually need a more refined one. I resonate with the risk @risicle sees, otherwise.

@dotlambda's proposal matches what I would do:

- If a package has both binary and source it will be marked as binary (i.e. we prefer a rougher schema, and err on the side of overapplying the "binary" label)
- it special cases bootstrap binaries and fonts
- Adds an `allowNonSourcePredicate`

I think this would be a great first step, and it would address the 2 points raised in the RFC:

- supply chain attacks
- patching

Other concerns, like wanting to only run free software + special casing some data files based on their license, I would leave out of this RFC, and potentially address in a follow-up one.